File: 3-leetcode/02-medium/1926-nearest-exit-from-entrance-in-maze.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
public:
    int nearestExit(vector<vector<char>> &maze, vector<int> &entrance)
    {
        queue<pair<int, int>> queue;
        queue.push({entrance[0], entrance[1]});

        int steps = 1;
        int rows = maze.size();
        int cols = maze[0].size();

        vector<vector<int>> dir = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

        maze[entrance[0]][entrance[1]] = '+';
        while (!queue.empty())
        {
            int l = queue.size();
            for (int k = 0; k < l; k++)
            {
                auto [i, j] = queue.front();
                queue.pop();

                for (int l = 0; l < 4; l++)
                {
                    int x = i + dir[l][0];
                    int y = j + dir[l][1];
                    if (x < 0 || y < 0 || x >= rows || y >= cols || maze[x][y] == '+') continue;
                    if (x == 0 || y == 0 || x == rows - 1 || y == cols - 1) return steps;
                    maze[x][y] = '+';
                    queue.push({x, y});
                }
            }
            steps++;
        }
        return -1;
    }
};

File: 3-leetcode/02-medium/js/1926-nearest-exit-from-entrance-in-maze.js
const nearestExit = function (maze, entrance) {
  let n = maze.length,
    m = maze[0].length;
  let steps = -1;
  let dir = [
    [-1, 0],
    [0, 1],
    [1, 0],
    [0, -1],
  ];
  let queue = [entrance];
  maze[entrance[0]][entrance[1]] = '+';

  while (queue.length) {
    steps++;
    let temp = queue.slice();
    queue = [];
    while (temp.length) {
      let [x, y] = temp.shift();
      if (steps && (x === 0 || x === n - 1 || y === 0 || y === m - 1))
        return steps;
      for (const [nx, ny] of dir) {
        let nextX = x + nx,
          nextY = y + ny;
        if (
          nextX >= 0 &&
          nextX < n &&
          nextY >= 0 &&
          nextY < m &&
          maze[nextX][nextY] !== '+'
        ) {
          maze[nextX][nextY] = '+';
          queue.push([nextX, nextY]);
        }
      }
    }
  }
  return -1;
};
